### PR TITLE
Pull Request for Compiler (All cases)

### PR DESCRIPTION
**Pull request by dylannakahodo**
Hey all, sorry this took so long but compiler is finally finished. 

This is a monster pull request, but the code for the compiler should be fairly straightforward as I based it off of the compiler from the paper.  The only big difference I believe is in the lambda case, I changed the next instruction from `(return)` to `(halt)` in the `(compile body next)`.  Troy and I decided this would be the best way to go about this, since there would be no guarantee that there would be a frame to return from.

**Please take a look at my unit tests to check if I constructed the expressions correctly.**
**Please also check if there are any mistakes or obvious edge cases I missed in my compiler implementation.**

Thanks!

### CODE REVIEW

File: src/shaka_scheme/system/vm/compiler/Compiler.cpp [hunk 1]
```diff
@@ -11,34 +11,135 @@ Compiler::Compiler() {}
...
+    Symbol& expression_type = car(input)->get<Symbol>();
+
+    // quote case
+    if (expression_type == Symbol("quote")) {
       Symbol instruction("constant");
       Data instruction_data(instruction);
 
-      DataPair next(*cdr(input), *list(create_node(*next_instruction)));
-      DataPair assembly(instruction_data, next);
+      return list(create_node(instruction_data), cdr(input), next_instruction);
```
Comment (btwooton) [inline src/shaka_scheme/system/vm/compiler/Compiler.cpp]: This seems dubious to me. If input is the list `(quote b)`, if I take the cdr of that, I should get back the list `(b)`, not the symbol 'b. However, I looked at your test case for compilation of `quote` and it appears to be working correctly, so I think this has to do with the questionable way in which the `cdr()` method is handling proper lists or pairs that represent proper lists. No biggie as long as the output is correct, but just want you to be aware that optically this looks wrong from a scheme semantics perspective. 
**Edit: See my comment on the unit-test for quote**
@dylannakahodo

File: tst/shaka_scheme/system/vm/unit-Compiler.cpp [hunk 2]
```diff
@@ -47,3 +48,219 @@ TEST(CompilerUnitTest, quote_test) {
   std::string output = ss.str();
```
Comment (btwooton) [inline tst/shaka_scheme/system/vm/unit-Compiler.cpp:50]: Did you verify that the input `constant_obj_pair` is actually the proper list `(quote "Hello")` and not the improper list (dotted-pair) `(quote . "Hello")`? If it happens to be the latter, that would explain the questionable call to `cdr()` used to retrieve the `obj_data` that I commented on above.  You can verify that it is properly formed by doing `std::cout << constant_obj_pair << std:endl` Please check this if you haven't already. @dylannakahodo

Comment (dylannakahodo) [reply tst/shaka_scheme/system/vm/unit-Compiler.cpp:50]: @Mortrax You're right. I was constructing it incorrectly, I'll fix this.

File: src/shaka_scheme/system/vm/compiler/Compiler.cpp (hunk 1 above)
Comment (dylannakahodo) [reply src/shaka_scheme/system/vm/compiler/Compiler.cpp]: @Mortrax So in this case, I should be calling `car(cdr(input))`?

Comment (btwooton) [reply src/shaka_scheme/system/vm/compiler/Compiler.cpp]: @dylannakahodo That is correct. If it is indeed the list `(quote <obj>)`, `obj` is at the `(car (cdr ..))` position. That should be the only change needed.

**Review by btwooton — CHANGES_REQUESTED** [review]
Well done Dylan. Barring the small changes to the compilation of `(quote ...)` that I mentioned in my comments, I can see from your implementation and unit-Tests that you have followed the specification of the compiler as laid out in Dybvig's dissertation to a T. As soon as you issue the commit fixing the compilation of quote and its associated unit test, I am ready to merge this pull request.

**Review by btwooton — APPROVED** [review]
Ah, just saw your commits. You're good to go!
